### PR TITLE
UCM: support setting trim and mmap thresholds via env variables

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

Take trim and mmap threshold values from
MALLOC_TRIM_THRESHOLD_
MALLOC_MMAP_THRESHOLD_

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4339/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4342/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4347/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
bot:bgate:retest

File: contrib/test_jenkins.sh [hunk 1]
```diff
@@ -466,6 +466,13 @@ run_gtest() {
 	$AFFINITY $TIMEOUT make -C test/gtest test
 	(cd test/gtest && rename .tap _gtest.tap *.tap && mv *.tap $GTEST_REPORT_DIR)
 
+    echo "==== Running malloc hooks mallopt() test ===="
+    # the test is very short. run it on every shard.
+    env UCX_IB_RCACHE=n UCX_GTEST_MALLOPT=yes \
+        GTEST_SHARD_INDEX=0 GTEST_TOTAL_SHARDS=1 \
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: why is this needed? it's enough 1 process will do the test

File: contrib/test_jenkins.sh [hunk 2]
```diff
@@ -466,6 +466,13 @@ run_gtest() {
...
+        GTEST_FILTER=malloc_hook_cplusplus.mallopt make -C test/gtest test
+	(cd test/gtest && rename .tap _gtest.tap *.tap && mv *.tap $GTEST_REPORT_DIR)
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: why is this needed?

File: src/ucm/malloc/malloc_hook.c [hunk 3]
```diff
@@ -600,6 +600,26 @@ static void ucm_malloc_install_optional_symbols()
...
+static void ucm_malloc_install_mallopt()
+{
+    /* copy values of M_MMAP_THRESHOLD and M_TRIM_THRESHOLD
+     * if they were overriden by the user
+     */
+    char *p;
+
+    p = getenv("MALLOC_TRIM_THRESHOLD_");
+    if (p) {
+        ucs_debug("set trim_thresh to %d", atoi(p));
```
Comment (yosefe) [inline src/ucm/malloc/malloc_hook.c]: i would use alol() and 'long' instead of atoi

File: test/gtest/ucm/malloc_hook.cc [hunk 4]
```diff
@@ -387,3 +387,43 @@ UCS_TEST_F(malloc_hook_cplusplus, new_delete) {
...
+UCS_TEST_F(malloc_hook_cplusplus, mallopt) {
+
+    int v;
+    char *p;
+
+    /* This test can not be run with the other
+     * tests because it assumes that malloc hooks
+     * are not initialized
+     */
+    if (getenv("UCX_GTEST_MALLOPT") == NULL) {
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: i would set MALLOC_TRIM_THRESHOLD_ in test_jenkins.sh, and inside this test - check if MALLOC_TRIM_THRESHOLD_ is present in getenv(). This way would not need to define another env var and no need to disable rcache.

File: test/gtest/ucm/malloc_hook.cc [hunk 5]
```diff
@@ -387,3 +387,43 @@ UCS_TEST_F(malloc_hook_cplusplus, new_delete) {
...
+
+    setenv("MALLOC_TRIM_THRESHOLD_", "12345", 1);
+    setenv("MALLOC_MMAP_THRESHOLD_", "425364", 1);
+
+    ucs_status_t result = ucm_set_event_handler(UCM_EVENT_VM_UNMAPPED,
+                                                0, mem_event_callback,
+                                                reinterpret_cast<void*>(this));
+    ASSERT_UCS_OK(result);
+
+    v = ucm_dlmallopt_get(M_TRIM_THRESHOLD);
```
Comment (yosefe) [inline test/gtest/ucm/malloc_hook.cc]: maybe better to do malloc+free and make sure we don't get an event notification?

File: contrib/test_jenkins.sh (hunk 1 above)
Comment (alex-mikheev) [reply contrib/test_jenkins.sh]: because gtest exits with non zero code if there are no tests to run

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (alex-mikheev) [reply contrib/test_jenkins.sh]: it seems to be a magic line that we do after ever gtest invocation

File: src/ucm/malloc/malloc_hook.c (hunk 3 above)
Comment (alex-mikheev) [reply src/ucm/malloc/malloc_hook.c]: mallopt() arguments are integers

File: test/gtest/ucm/malloc_hook.cc (hunk 4 above)
Comment (alex-mikheev) [reply test/gtest/ucm/malloc_hook.cc]: you still have to disable rcache because it triggers malloc initialization way before the actual test

File: contrib/test_jenkins.sh (hunk 1 above)
Comment (yosefe) [reply contrib/test_jenkins.sh]: ok

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (yosefe) [reply contrib/test_jenkins.sh]: i think this can override the previous tap files (of the main test - can you please check it)?

File: src/ucm/malloc/malloc_hook.c (hunk 3 above)
Comment (yosefe) [reply src/ucm/malloc/malloc_hook.c]: ok

File: test/gtest/ucm/malloc_hook.cc (hunk 4 above)
Comment (yosefe) [reply test/gtest/ucm/malloc_hook.cc]: ok, but still can avoid UCX_GTEST_MALLOPT

File: contrib/test_jenkins.sh (hunk 2 above)
Comment (alex-mikheev) [reply contrib/test_jenkins.sh]: I see that only malloc_hook_cplusplus.tap is created

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4365/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4470/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4478/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4479/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4487/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4493/ for details (Mellanox internal link).

**Comment by miked-mellanox** [comment]
+1
